crates/lair_keystore_api/src/lair_server/priv_api.rs

```diff
@@ -799,39 +920,54 @@ pub(crate) fn priv_req_new_wka_tls_cert<'a>(
 }
 
 pub(crate) async fn derive_ed(
-    seed: &BufReadSized<32>,
-) -> LairResult<(Ed25519PubKey, BufReadSized<64>)> {
-    // get the signature keypair
-    let ed_pk = sodoken::BufWriteSized::new_no_lock();
-    let ed_sk = sodoken::BufWriteSized::new_mem_locked()?;
-    sodoken::sign::seed_keypair(ed_pk.clone(), ed_sk.clone(), seed.clone())
-        .await?;
+    seed: SharedSizedLockedArray<32>,
+) -> LairResult<(Ed25519PubKey, sodoken::SizedLockedArray<64>)> {
+    let (ed_pk, ed_sk) =
+        tokio::task::spawn_blocking(move || -> LairResult<_> {
+            // get the signature keypair
+            let mut ed_pk = [0; sodoken::sign::PUBLICKEYBYTES];
+            let mut ed_sk = sodoken::SizedLockedArray::<
+                { sodoken::sign::SECRETKEYBYTES },
+            >::new()?;
+
+            sodoken::sign::seed_keypair(
+                &mut ed_pk,
+                &mut ed_sk.lock(),
+                &seed.lock().lock(),
+            )?;
 
-    let ed_pk: Ed25519PubKey = ed_pk.try_unwrap_sized().unwrap().into();
-    let ed_sk = ed_sk.to_read_sized();
+            Ok((ed_pk, ed_sk))
+        })
+        .await
+        .map_err(OneErr::new)??;
 
-    Ok((ed_pk, ed_sk))
+    Ok((ed_pk.into(), ed_sk))
 }
 
 pub(crate) async fn derive_x(
-    seed: &BufReadSized<32>,
-) -> LairResult<(X25519PubKey, BufReadSized<32>)> {
-    // get the encryption keypair
-    let x_pk = sodoken::BufWriteSized::new_no_lock();
-    let x_sk = sodoken::BufWriteSized::new_mem_locked()?;
-    // we're using the chacha sodium keypair api here, but the
-    // keypairs are valid also for salsa.
-    sodoken::crypto_box::curve25519xchacha20poly1305::seed_keypair(
-        x_pk.clone(),
-        x_sk.clone(),
-        seed.clone(),
-    )
-    .await?;
-
-    let x_pk: X25519PubKey = x_pk.try_unwrap_sized().unwrap().into();
-    let x_sk = x_sk.to_read_sized();
-
-    Ok((x_pk, x_sk))
+    seed: SharedSizedLockedArray<32>,
+) -> LairResult<(X25519PubKey, sodoken::SizedLockedArray<32>)> {
+    let (x_pk, x_sk) = tokio::task::spawn_blocking(move || -> LairResult<_> {
+        // get the encryption keypair
+        let mut x_pk = [0; sodoken::crypto_box::XSALSA_PUBLICKEYBYTES];
+        let mut x_sk = sodoken::SizedLockedArray::<
+            { sodoken::crypto_box::XSALSA_SECRETKEYBYTES },
+        >::new()?;
+
+        // we're using the chacha sodium keypair api here, but the
+        // keypairs are valid also for salsa.
```

Comment:
```suggestion
```

haha, now we're using the salsa one

Reply:
https://github.com/holochain/lair/pull/143/commits/b6f64ac499f6c25e33b6b125e4b3e59e90761dca